controllers/persistentvolumeclaim_controller.go

```diff
@@ -0,0 +1,260 @@
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+var (
+	rsCronJobScheduleTimeAnnotation = "reclaimspace." + csiaddonsv1alpha1.GroupVersion.Group + "/schedule"
+	rsCronJobNameAnnotation         = "reclaimspace." + csiaddonsv1alpha1.GroupVersion.Group + "/cronjob"
+)
+
+const (
+	defaultSchedule = "@weekly"
+)
+
+//+kubebuilder:rbac:groups=core,resources=persistentvolumeclaims,verbs=get;list;watch;patch
+//+kubebuilder:rbac:groups=csiaddons.openshift.io,resources=reclaimspacecronjobs,verbs=get;list;watch;create;delete
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// This is triggered when `reclaimspace.csiaddons.openshift/schedule` annotation
+// is found on newly created PVC or if there is a change in valude of the annotation.
+func (r *PersistentVolumeClaimReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	// Fetch PersistentVolumeClaim instance
+	pvc := &corev1.PersistentVolumeClaim{}
+	err := r.Client.Get(ctx, req.NamespacedName, pvc)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			logger.Info("PersistentVolumeClaim resource not found")
+
+			return ctrl.Result{}, nil
+		}
+
+		return ctrl.Result{}, err
+	}
+
+	// delete child cronjob.
+	err = r.deleteChildCronJob(ctx, &logger, &req)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	annotations := pvc.GetAnnotations()
+	schedule, scheduleFound := getScheduleFromAnnotation(&logger, annotations)
+	if !scheduleFound {
+		// if schedule is not found and name is found, delete name from annotation.
+		_, nameFound := annotations[rsCronJobNameAnnotation]
+		if nameFound {
+			// remove name annotation by patching it to null.
+			patch := []byte(fmt.Sprintf(`{"metadata":{"annotations":{%q: null}}}`, rsCronJobNameAnnotation))
+			err = r.Client.Patch(ctx, pvc, client.RawPatch(types.StrategicMergePatchType, patch))
+			if err != nil {
+				logger.Error(err, "Failed to update annotation")
+
+				return ctrl.Result{}, err
+			}
+		}
+		logger.Info("Annotation not set, exiting reconcile")
+		// no schedule annotation set, just dequeue.
+		return ctrl.Result{}, nil
+	}
+
+	// create cronjob name
+	rsCronJobName := fmt.Sprintf("%s-%v", pvc.Name, time.Now().Unix())
+	rsCronJob := constructRSCronJob(rsCronJobName, req.Namespace, schedule, pvc.Name)
+	err = ctrl.SetControllerReference(pvc, rsCronJob, r.Scheme)
+	if err != nil {
+		logger.Error(err, "Failed to set controllerReference")
+
+		return ctrl.Result{}, err
+	}
+
+	err = r.Client.Create(ctx, rsCronJob)
+	if err != nil {
+		logger.Error(err, "Failed to create reclaimSpaceCronJob")
+
+		return ctrl.Result{}, err
+	}
+
+	// add cronjob name in annotations.
+	patch := []byte(fmt.Sprintf(`{"metadata":{"annotations":{%q:%q}}}`, rsCronJobNameAnnotation, rsCronJobName))
```

Comment:
may be name it annotations instead of patch?

Reply:
> may be name it annotations instead of patch?

annotations is already taken, just patch sounds fine to me